packages/core/src/electron-main/messaging/electron-main-messaging-service.ts

```diff
@@ -0,0 +1,120 @@
+import { createWebSocketConnection } from 'vscode-ws-jsonrpc/lib/socket/connection';
+import { ContributionProvider } from '../../common/contribution-provider';
+import { WebSocketChannel } from '../../common/messaging/web-socket-channel';
+import { ConsoleLogger } from '../../node/messaging/logger';
+import { ElectronMainConnectionHandler, ElectronBackendMessage, ElectronBackendConnectionPipe } from '../../electron-common/messaging/electron-backend-connection-handler';
+import { MessagingContribution } from '../../node/messaging/messaging-contribution';
+import { ChildProcess } from 'child_process';
+
+export interface ElectronMainConnectionOptions {
+}
+
+/**
+ * This component replicates the role filled by `MessagingContribution` but for connecting the backend with the electron process.
+ * It is based on process communication using the created backend process.
+ * Alternatively it uses a simple message pipe object when running in `--no-cluster` mode.
+ *
+ * This component allows communication between server process (backend) and electron main process.
+ */
+@injectable()
+export class ElectronMainMessagingService {
+
+    @inject(ContributionProvider) @named(ElectronMainConnectionHandler)
+    protected readonly connectionHandlers: ContributionProvider<ElectronMainConnectionHandler>;
+
+    protected readonly channelHandlers = new MessagingContribution.ConnectionHandlers<WebSocketChannel>();
+    protected readonly channels = new Map<number, WebSocketChannel>();
+    protected backendProcess?: ChildProcess;
+
+    start(backendProcess?: ChildProcess): void {
+        if (backendProcess) {
+            this.backendProcess = backendProcess;
+            this.backendProcess.on('message', message => {
+                if (ElectronBackendMessage.is(message)) {
+                    this.handleMessage(ElectronBackendMessage.get(message));
+                }
+            });
+            this.backendProcess.on('exit', () => {
+                this.closeChannels();
+            });
+        } else {
+            ElectronBackendConnectionPipe.onMessage('electron', message => {
+                this.handleMessage(message);
+            });
+        }
+        for (const connectionHandler of this.connectionHandlers.getContributions()) {
+            this.channelHandlers.push(connectionHandler.path, (params, channel) => {
+                const connection = createWebSocketConnection(channel, new ConsoleLogger());
+                connectionHandler.onConnection(connection);
+            });
+        }
+    }
+
+    protected closeChannels(): void {
+        for (const channel of Array.from(this.channels.values())) {
+            channel.close(undefined, 'Backend exited');
+        }
+        this.channels.clear();
+    }
+
+    protected handleMessage(data: string): void {
+        try {
+            // Start parsing the message to extract the channel id and route
+            const message: WebSocketChannel.Message = JSON.parse(data.toString());
```

Comment:
One point of the discussion re: performance in our communication layer is that conversions back and forth from string to JSON are relatively expensive. Will all messages be passing through here, or only some, or how will it fit into the refactor of the messaging layer currently underway? Is it expected that this will be one terminus of the messaging system that can delegate the parsed messages to handlers, or will further parsing be necessary?